src/webgpu/api/operation/render_pipeline/sample_mask.spec.ts

```diff
@@ -1,16 +1,443 @@
+
+    if (
+      // rasterization needs to cover the center of the pixel
+      (rasterizationMask & 15) >= 15 &&
+      // There is only one sample
+      (fragmentShaderOutputMask & sampleMask & 1) !== 0
+    ) {
+      // When full screen quad is drawn, Texel 3 is sampled at the pixel center
+      expected = kColors[3];
+    }
+
+    this.expectSingleColor(texture, format, {
+      size: [1, 1, 1],
+      exp: {
+        R: expected.R / 0xff,
+        G: expected.G / 0xff,
+        B: expected.B / 0xff,
+        A: expected.A / 0xff,
+      },
+    });
+  }
+
+  CheckMultiSampledResult(
+    texture: GPUTexture,
+    sampleCount: number,
+    rasterizationMask: number,
+    sampleMask: number,
+    fragmentShaderOutputMask: number
+  ) {
+    const expected = {
+      R: 0,
+      G: 0,
+      B: 0,
+      A: 0,
+    };
+    range(4, (id: number) => {
+      const m = rasterizationMask & sampleMask & fragmentShaderOutputMask & (1 << id);
+      // WebGPU only support up to 4 samples, so samples after the first 4 should be ignored.
+      const s = (m & 0xf) === 0 ? kEmptySample : kColors[id % 4];
+      expected.R += s.R;
+      expected.G += s.G;
+      expected.B += s.B;
+      expected.A += s.A;
+    });
+
+    this.expectSinglePixelBetweenTwoValuesIn2DTexture(
+      texture,
+      format,
+      { x: 0, y: 0 },
+      {
+        exp: [
+          new Uint8Array([
+            Math.floor(expected.R / sampleCount),
+            Math.floor(expected.G / sampleCount),
+            Math.floor(expected.B / sampleCount),
+            Math.floor(expected.A / sampleCount),
+          ]),
+          new Uint8Array([
+            Math.ceil(expected.R / sampleCount),
+            Math.ceil(expected.G / sampleCount),
+            Math.ceil(expected.B / sampleCount),
+            Math.ceil(expected.A / sampleCount),
+          ]),
+        ],
+      }
+    );
+  }
+}
+
+export const g = makeTestGroup(F);
+
+g.test('final_output')
+  .desc(
+    `
+Tests that the final sample mask is the logical AND of all the relevant masks -- meaning that the samples
+not included in the final mask are discarded on any attachments including
+- color outputs
 TODO:
+- depth tests
+- stencil operations
+
+The test draws 0/1/1+ textured quads of which each sample in the standard 4-sample pattern results in a different color:
+- Sample 0, Texel 0, top-left: Red
+- Sample 1, Texel 1, top-left: Green
+- Sample 2, Texel 2, top-left: Blue
+- Sample 3, Texel 3, top-left: Yellow
+
+The test checks which sample is passed by looking at the final color of the 1x1 resolve target texture, to see if that matches
+what is expected given by the rasterization mask, sample mask, and fragment shader output mask.
+
 - for sampleCount = { 1, 4 } and various combinations of:
-    - rasterization mask = { 0, 1, 2, 3, 15 }
+    - rasterization mask = { 0, 1, 2, 3, 4, 5, 15 }
     - sample mask = { 0, 1, 2, 3, 15, 30 }
     - fragment shader output mask (SV_Coverage) = { 0, 1, 2, 3, 15, 30 }
```

Comment:
Can update the language here, this description dates back to when we thought we might do WHLSL.

```suggestion
    - fragment shader output @builtin(sample_mask) = { 0, 1, 2, 3, 15, 30 }
```